lib/rack/idempotency.rb

```diff
@@ -19,23 +20,47 @@ class Idempotency
     def initialize(app, store: NullStore.new)
       @app     = app
       @store   = store
+      if @store.class.to_s == "Rack::Idempotency::RedisStore"
+        @mutex_mode = true
+        @cache_errors = true
+      end
     end
 
     def call(env)
       request = Request.new(env.dup.freeze)
-      storage = RequestStorage.new(@store, request)
+      read_response(request, env).to_a
+    end
 
-      storage.read || store_response(storage, env)
+    def read_response(request, env)
+      if request.idempotency_key
+        @storage = RequestStorage.new(@store, request)
+        @storage.read || store_response(env)
+      else
+        Response.new(*@app.call(env))
+      end
     end
 
     private
 
-    def store_response(storage, env)
+    def fetch_and_cache(env)
       response = Response.new(*@app.call(env))
+      @storage.write(response) if @cache_errors || response.success?
+      response
+    end
 
-      storage.write(response) if response.success?
```

Comment:
Oh wow, big bug in the original gem! It only stores only `.success?` responses (so only `2XX` status codes). What about redirects? Granted, usually APIs don't respond with redirects, but still, this looks like a big oversight.